File: src/main/java/io/funky/fangs/keep_it_personal/KeepItPersonalModInitializer.java
package io.funky.fangs.keep_it_personal;

import io.funky.fangs.keep_it_personal.command.*;
import io.funky.fangs.keep_it_personal.configuration.KeepItPersonalConfiguration;
import io.funky.fangs.keep_it_personal.domain.DeathPreference;
import io.funky.fangs.keep_it_personal.exception.KeepItPersonalException;
import me.lucko.fabric.api.permissions.v0.Permissions;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.command.v2.CommandRegistrationCallback;
import net.minecraft.server.command.CommandManager;

import java.util.Arrays;

import static com.mojang.brigadier.arguments.BoolArgumentType.bool;

public class KeepItPersonalModInitializer implements ModInitializer {
    public static final String MOD_ID = "keep_it_personal";

    @Override
    public void onInitialize() {
        final var configuration = KeepItPersonalConfiguration.getInstance();
        final var preferences = configuration.preferences();
        final var disabled = preferences.disabled();
        final var enabled = preferences.enabled();
        final int permissionLevel = configuration.permissions().permissionLevel();

        final var hasIntersectingPreferences = Arrays.stream(DeathPreference.values())
                .anyMatch(deathPreference -> enabled.contains(deathPreference)
                        && disabled.contains(deathPreference));

        if (hasIntersectingPreferences) {
            throw new KeepItPersonalException("Overlapping preferences in disabled and enabled preferences are not permitted");
        }
        else if (permissionLevel < 0 || permissionLevel > 4) {
            throw new KeepItPersonalException("Permission level must be between 0 and 4");
        }

        CommandRegistrationCallback.EVENT.register((dispatcher, access, environment) -> {
            final var kipCommand = CommandManager.literal(KeepItPersonalCommand.NAME);

            dispatcher.register(
                    kipCommand.requires(Permissions.require(KeepItPersonalCommand.PERMISSION, permissionLevel))
                            .executes(new ViewDeathPreferencesCommand())
            );

            for (final var deathPreference : DeathPreference.values()) {
                if (!enabled.contains(deathPreference) && !disabled.contains(deathPreference)) {
                    final var permission = KeepItPersonalCommand.PERMISSION + '.' + deathPreference;
                    dispatcher.register(
                            kipCommand.then(CommandManager.literal(deathPreference.toString())
                                    .requires(Permissions.require(permission, permissionLevel))
                                    .executes(new GetDeathPreferenceCommand(deathPreference))
                                    .then(CommandManager.argument(ModifyDeathPreferenceCommand.PARAMETER, bool())
                                            .requires(Permissions.require(permission, permissionLevel))
                                            .executes(new ModifyDeathPreferenceCommand(deathPreference))
                                    )
                            )
                    );
                }
            }

            dispatcher.register(
                    kipCommand.then(CommandManager.literal(FillDeathPreferencesCommand.NAME)
                            .requires(Permissions.require(FillDeathPreferencesCommand.PERMISSION, permissionLevel))
                            .executes(new FillDeathPreferencesCommand())
                    )
            );

            dispatcher.register(
                    kipCommand.then(CommandManager.literal(ClearDeathPreferencesCommand.NAME)
                            .requires(Permissions.require(ClearDeathPreferencesCommand.PERMISSION, permissionLevel))
                            .executes(new ClearDeathPreferencesCommand())
                    )
            );
        });
    }
}

File: src/main/kotlin/io/funky/fangs/keep_it_personal/KeepItPersonalModInitializer.kt
package io.funky.fangs.keep_it_personal

import com.fasterxml.jackson.dataformat.toml.TomlMapper
import com.mojang.brigadier.arguments.BoolArgumentType
import io.funky.fangs.keep_it_personal.command.*
import io.funky.fangs.keep_it_personal.configuration.KeepItPersonalConfiguration
import me.lucko.fabric.api.permissions.v0.Permissions
import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.command.v2.CommandRegistrationCallback
import net.fabricmc.loader.api.FabricLoader
import net.minecraft.server.command.CommandManager

class KeepItPersonalModInitializer: ModInitializer {
    companion object {
        const val MOD_ID = "keep_it_personal"

        private val CONFIGURATION_FILE = FabricLoader.getInstance().configDir.resolve("$MOD_ID.toml").toFile()
        val CONFIGURATION: KeepItPersonalConfiguration by lazy {
            TomlMapper().readerFor(KeepItPersonalConfiguration::class.java).readValue(CONFIGURATION_FILE)
        }
    }

    @Override
    override fun onInitialize() {
        if (!CONFIGURATION_FILE.exists()) {
            CONFIGURATION_FILE.createNewFile()
        }

        val (disabled, enabled) = CONFIGURATION.preferences
        if (enabled.intersect(disabled).isNotEmpty()) {
            throw IllegalArgumentException("Overlapping preferences in disabled and enabled are not permitted")
        }

        CommandRegistrationCallback.EVENT.register { dispatcher, _, _ ->
            val keepCommand = CommandManager.literal(KEEP_COMMAND)

            val defaultPermissionLevel = CONFIGURATION.permissions.permissionLevel

            dispatcher.register(
                keepCommand.requires(Permissions.require("${MOD_ID}.${KEEP_COMMAND}", defaultPermissionLevel))
                    .executes(ViewDeathPreferencesCommand())
            )

            DeathPreference.entries
                .filterNot { enabled.contains(it) || disabled.contains(it) }
                .forEach {
                    dispatcher.register(
                        keepCommand.then(CommandManager.literal(it.toString())
                            .requires(Permissions.require(
                                "${MOD_ID}.${KEEP_COMMAND}.${it}",
                                defaultPermissionLevel
                            ))
                            .executes(GetDeathPreferenceCommand(it))
                            .then(CommandManager.argument(ARGUMENT_NAME, BoolArgumentType.bool())
                                .requires(Permissions.require(
                                    "${MOD_ID}.${KEEP_COMMAND}.${it}",
                                    defaultPermissionLevel
                                ))
                                .executes(SetDeathPreferenceCommand(it))
                            )
                        )
                    )
                }

            dispatcher.register(
                keepCommand.then(CommandManager.literal(FillDeathPreferencesCommand.NAME)
                    .requires(Permissions.require(
                        "${MOD_ID}.${KEEP_COMMAND}.${FillDeathPreferencesCommand.NAME}",
                        defaultPermissionLevel
                    ))
                    .executes(FillDeathPreferencesCommand())
                )
            )

            dispatcher.register(
                keepCommand.then(CommandManager.literal(ClearDeathPreferencesCommand.NAME)
                    .requires(Permissions.require(
                        "${MOD_ID}.${KEEP_COMMAND}.${ClearDeathPreferencesCommand.NAME}",
                        defaultPermissionLevel
                    ))
                    .executes(ClearDeathPreferencesCommand())
                )
            )
        }
    }
}

